docs/source/format/CProtocol.rst

```diff
@@ -0,0 +1,560 @@
+
+   Mandatory.  A null-terminated, UTF8-encoded string describing
+   the data type.  If the data type is nested, child types are not
+   encoded here but in the :c:member:`ArrowArray.children` arrays.
+
+   Consumers MAY decide not to support all data types, but they
+   should document this limitation.
+
+.. c:member:: const char* ArrowArray.name
+
+   Optional.  A null-terminated, UTF8-encoded string of the field
+   or array name.  This is mainly used to reconstruct child fields
+   of nested arrays.
+
+   Producers MAY decide not to provide this information, and consumers
+   MAY decide to ignore it.  If omitted, MAY be NULL or an empty string.
+
+.. c:member:: const char* ArrowArray.metadata
+
+   Optional.  A null-terminated, UTF8-encoded string describing
+   the type's metadata.  If the data type is nested, child types are not
+   encoded here but in the :c:member:`ArrowArray.children` arrays.
+
+   MUST be a JSON-compatible mapping of UTF8 strings to UTF8 strings.
+   Whitespace MUST only use the 0x20 character. Example::
+
+      {"key1": "base64-encoded value1", "key2": "base64-encode value2"}
+
+   If omitted, MUST be NULL (not an empty string).
+
+   Consumers MAY choose to ignore metadata.  Even then, they SHOULD keep
+   the metadata string around so as to propagate its information to their
+   own consumers.
+
+.. c:member:: int64_t ArrowArray.flags
+
+   Optional.  A bitfield of flags enriching the type or array description.
+   Its value is computed by OR'ing together the flag values.
+   The following flags are available:
+
+   * ``ARROW_FLAG_NULLABLE``: whether this field is semantically nullable
+     (regardless of whether it actually has null values).
+   * ``ARROW_FLAG_ORDERED``: for dictionary-encoded arrays, whether the
+     ordering of dictionary indices is semantically meaningful.
+
+   If omitted, MUST be 0.
+
+   Consumers MAY choose to ignore some or all of the flags.  Even then,
+   they SHOULD keep this value around so as to propagate its information
+   to their own consumers.
+
+.. c:member:: int64_t ArrowArray.length
+
+   Mandatory.  The logical length of the array (i.e. its number of items).
+
+.. c:member:: int64_t ArrowArray.null_count
+
+   Mandatory.  The number of null items in the array.  MAY be -1 if not
+   yet computed.
+
+.. c:member:: int64_t ArrowArray.offset
+
+   Mandatory.  The logical offset inside the array (i.e. the number of items
+   from the physical start of the buffers).  MUST be 0 or positive.
+
+   Producers MAY specify that they will only produce 0-offset arrays to
+   ease implementation of consumer code.
+   Consumers MAY decide not to support non-0-offset arrays, but they
+   should document this limitation.
+
+.. c:member:: int64_t ArrowArray.n_buffers
+
+   Mandatory.  The number of physical buffers backing this array.  The
+   number of buffers is a function of the data type, as described in the
+   :ref:`Columnar format specification <format_columnar>`.
```

Comment:
I am not sure if this problem should be discussed here: is it possible for a null value to hold non-empty space in the data buffer?
There is a similar discussion for the Java code base:
https://www.mail-archive.com/dev@arrow.apache.org/msg14284.html

Reply:
I don't think it should be discussed here, no. I'm a bit surprised by Jacques' answer in that discussion, though. I would have said "1. a null value always takes no space".

Reply:
@liyafan82 I'm a bit confused by your question here. I don't see how it's related to the List format discussion. We have taken the position uniformly that we do not specify what data is "underneath" a null. So in a ListVector a non-empty element could be "made null" by applying a different validity bitmap. To do otherwise would be inconsistent

Reply:
@pitrou Thanks for your clarification.
@wesm I think that discussion is also related to the layout of ListVector. In particular, whether a null in a list can have some position in the underlying vector is similar to the problem of whether a null value in VarCharVector can hold some space in the data buffer.